docs/site/Life-cycle.md

```diff
@@ -0,0 +1,217 @@
+  - stop: Custom logic to be invoked when the application stops
+
+## The `LifeCycleObserver` interface
+
+To react on life cycle events, a life cycle observer implements the
+`LifeCycleObserver` interface.
+
+```ts
+import {ValueOrPromise} from '@loopback/context';
+
+/**
+ * Observers to handle life cycle start/stop events
+ */
+export interface LifeCycleObserver {
+  start?(): ValueOrPromise<void>;
+  stop?(): ValueOrPromise<void>;
+}
+```
+
+Please note all methods are optional so that an observer can opt in certain
+events. Each main events such as `start` and `stop` are further divided into
+three sub-phases to allow the multiple-step processing.
+
+## Register a life cycle observer
+
+A life cycle observer can be registered by calling `lifeCycleObserver()` of the
+application. It binds the observer to the application context with a special
+tag - `CoreTags.LIFE_CYCLE_OBSERVER`.
+
+```ts
+app.lifeCycleObserver(MyObserver);
+```
+
+Please note that `app.server()` automatically registers servers as life cycle
+observers.
+
+Life cycle observers can be registered via a component too:
+
+```ts
+export class MyComponentWithObservers {
+  lifeCycleObservers = [XObserver, YObserver];
+}
+```
+
+## Discover life cycle observers
+
+The `Application` finds all bindings tagged with `CoreTags.LIFE_CYCLE_OBSERVER`
+within the context chain and resolve them as observers to be notified.
+
+## Notify life cycle observers of start/stop related events by order
+
+There may be dependencies between life cycle observers and their order of
+processing for `start` and `stop` need to be coordinated. For example, we
+usually start a server to listen on incoming requests only after other parts of
+the application are ready to handle requests. The stop sequence is typically
+processed in the reverse order. To support such cases, we introduce
+two-dimension steps to control the order of life cycle actions.
+
+### Observer groups
+
+First of all, we allow each of the life cycle observers to be tagged with a
+group. For example:
+
+- datasource (connect/disconnect)
+
+  - mongodb
+  - mysql
+
+- server
+  - rest
+  - gRPC
+
+We can then configure the application to trigger observers group by group as
+configured by an array of groups in order such as `['datasource', 'server']`.
+Observers within the same group can be notified in parallel.
+
+For example,
+
+```ts
+app
+  .bind('observers.MyObserver')
+  .toClass(MyObserver)
+  .tag({
+    [CoreTags.LIFE_CYCLE_OBSERVER_GROUP]: 'g1',
+  })
+  .apply(asLifeCycleObserver);
+```
+
+The observer class can also be decorated with `@bind` to provide binding
+metadata.
+
+```ts
+import {bind, createBindingFromClass} from '@loopback/context';
+import {CoreTags, asLifeCycleObserver} from '@loopback/core';
+
+@bind(
+  {
+    tags: {
+      [CoreTags.LIFE_CYCLE_OBSERVER_GROUP]: 'g1',
+    },
+  },
+  asLifeCycleObserver,
+)
+export class MyObserver {
+  // ...
+}
+
+app.add(createBindingFromClass(MyObserver));
```

Comment:
Probably out of scope of this pull request: would it make sense to introduce Context-level sugar API for adding a binding created from a class?

```diff
- app.add(createBindingFromClass(MyObserver));
+ app.bindFromClass(MyObserver);
```

Reply:
+1.